Fix #3935 OSErrors being hidden

Raymo111 opened the pull request:
Fixes #3935

## Explanations
- `FSTests:test_srcdir_duplicate` -> `FS.Dir:bld1.srcdir_duplicate('exists')` -> `FS.File:node.do_duplicate(srcnode)` -> `FS:Unlink(self, None, None)` -> `FS:UnlinkFunc` -> `t.fs.unlink(file)` -> throws `FileNotFoundError` when file does not exist, wrap in try-except
- The new fix means `e = Link(self, src, None)` in `FS.File:node.do_duplicate()` raises `SCons.Errors.BuildError` instead of returning them, so we catch it and raise (along with raising when an error is returned)

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [x] I have updated the appropriate documentation

mwichmann commented:
That led to rather a lot of test failures.  Yes, there was indeed a period when there was a problem with the Python `EnvironmentError` possibly overlapping with the SCons `EnvironmentError`; the latter was renamed to reduce problem potential. But this tweak seems to be breaking something.

Raymo111 commented:
@mwichmann yes, this tweak correctly breaks a bunch tests that appeared to be fine but were actually throwing filesystem errors (OSErrors) that were being incorrectly suppressed.

mwichmann commented:
Haven't had a chance to dig in any detail, but a couple of tests I looked at were negative cases - we do something, expect a file *not* to have been created, and now the must_not_exist method is failing?  I'll get to this at some point - yes, I realize now that some of this comes from a comment I made on the linked bug a long time ago, but it still seems an unexpected result.

Raymo111 commented:
@mwichmann I fixed most of the tests, and updated the PR description above with explanations of the changes. There are 3 more failing tests that I don't really know how to fix:

- After the fix, test/Copy-Symlink.py correctly errors out at `Execute( Copy( 'Fails', '%(badlinkToCopy)s', False ) )`. Not sure what to do with the rest of the test

I'm not sure where exactly the other two tests are failing.

- I know `test/Execute.py` is failing due to a `nonexistent.in: No such file or directory` not expected by the test from `Execute(Copy('nonexistent.out', 'nonexistent.in'))`.
- I know `test/VariantDir/errors.py` is failing because the readonly SConsCript (`test.run(chdir = 'ro-SConscript'`) is failing before it gets to `raise SCons.Errors.StopError` in `FS.py:do_duplicate()`.

I would love some help/pointers to how to fix/change these three tests!

bdbaddog commented:
@Raymo111 - did you fix the tests? or fix in code? I don't see any test updated pushed to this PR?

Raymo111 commented:
I fixed the code to make the tests work. The tests I fixed weren't broken, the code was.

mwichmann commented:
>  `test/Execute.py` is failing due to a `nonexistent.in: No such file or directory` not expected by the test from `Execute(Copy('nonexistent.out', 'nonexistent.in'))`.

I don't think that's exactly the failure, that message is very clearly expected (conditionally, since Windows emits a different bit of text). One way you can debug a bit more is to run like this:
```bash
PRESERVE=1 ./runtest.py test/Execute.py
```
That will avoid deleting the test directory, and at the end it will tell you the path to it (twice, for some reason).  You can then go look around in there manually.

mwichmann commented:
Okay, revise the previous comment. The `no such file or directory` message is now appearing on stdout, where it was expected on stderr previously.

mwichmann commented:
As far as the last one - you write:

> After the fix, `test/Copy-Symlink.py` correctly errors out at `Execute( Copy( 'Fails', '%(badlinkToCopy)s', False ) )`.

But in fact, erroring out here is *not* correct.  From the documentation (emphasis mine):

> `Execute` returns the exit value of the command or return value of the Python function. scons prints an error message if the executed action fails (exits with or returns a non-zero value), however *<ins>it does not, automatically terminate the build for such a failure</ins>*. If you want the build to stop in response to a failed Execute call, you must explicitly check for a non-zero return value:

Raymo111 commented:
> Messages expected on stderr have moved to stdout

Which behavior is correct? Should these OSErrors be displayed on stderr or stdout?

> But in fact, erroring out here is *not* correct

Hmm ok, but in that case then the result should never be `raise`ed in `FunctionAction:execute()` should it? Currently all errors except `EnvironmentError`s are raised instead of returned, and after this PR all errors except `SConsEnvionmentError`s will be raised instead of returned. Doesn't this mean that the whole `FIXME raise result` here should be removed?

mwichmann commented:
> > Messages expected on stderr have moved to stdout
> 
> Which behavior is correct? Should these OSErrors be displayed on stderr or stdout?
> 
> > But in fact, erroring out here is _not_ correct
> 
> Hmm ok, but in that case then the result should never be `raise`ed in `FunctionAction:execute()` should it? Currently all errors except `EnvironmentError`s are raised instead of returned, and after this PR all errors except `SConsEnvionmentError`s will be raised instead of returned. Doesn't this mean that the whole `FIXME raise result` here should be removed?

So I'm going to make a random comment that's not going to help:  after more than seven years working with/on SCons, I'm still discovering little gems where "somebody thought that was a good idea 20 years ago" except it's not necessarily entirely logical.  Don't know the answers to your questions... it seems like this `Execute` having a "special case" where it doesn't actually fail on a failure is one of those.

Raymo111 commented:
@mwichmann Trying something new... I've removed the raising exceptions exception entirely for more consistent behavior, could you approve workflow please?

bdbaddog commented:
Minor nit. Blurbs in CHANGES.txt should be sorted by contributors last name, so yours should be before @mwichmann 's. Please fix?

Raymo111 commented:
@bdbaddog done.
